### PR TITLE
Enabled masked scatter through decomposition

### PR DESCRIPTION
**Pull request by shruthip-venkatesh**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Implements aten.masked_scatter for the Spyre backend by decomposing it into operations the hardware can already express, and fixes a silent LX scratchpad corruption that surfaces when an ExternKernel (e.g. a CPU-fallback bool→fp32 cast) is scheduled between a producer and consumer that share an LX-resident buffer.

#### Which issue(s) this PR is related to:

https://github.com/torch-spyre/torch-spyre/issues/2637

#### Additional note:

The test uses a fixed seed (torch.manual_seed(0)) and exactly N_TRUE = SRC_ROWS = 266 selected rows so the source is never over-consumed. Inputs are round-tripped through the device before building the CPU reference to cancel SEN169_FP16 mantissa rounding (9-bit vs IEEE 10-bit), making the comparison an exact value check rather than a tolerance check.

### CODE REVIEW

File: torch_spyre/_inductor/decompositions.py [hunk 1]
```diff
@@ -997,6 +997,84 @@ def spyre_prod_dim_int(
...
+    # The gather stays 2D (its args are indirect, so the pointwise dim_order
+    # projection skips them), but the `where` must stay at `self`'s rank:
+    # that projection computes `rank_diff = len(output) - len(arg)` and only
+    # handles inputs of *lower* rank. A lower-rank output against the full-rank
+    # ND mask gives rank_diff < 0, which shifts dims the wrong way and builds a
+    # layout whose dim_order rank no longer matches host_size ("Incompatible
+    # host_size and dim_order"). Reshaping back to ND keeps every `where`
+    # input rank-aligned with the output.
+    gathered = source_2d[row_idx].reshape(self.shape)
+    return torch.where(mask, gathered, self)
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/decompositions.py:1044]: Crash on statically-empty source with an all-False mask. When source_2d.shape[0] == 0 (a legal masked_scatter call: empty source paired with an all-False mask — the source only needs to be at least as large as the True-count, which is 0 here), every unselected row still computes row_idx = 0 and source_2d[row_idx] indexes into a dim-0 of size 0, raising an IndexError. source_2d.shape[0] is a static shape value known at trace time, so this can be guarded without any data-dependent branching.

File: torch_spyre/_inductor/scratchpad/allocator.py
Comment (Vikram123-byte) [inline torch_spyre/_inductor/scratchpad/allocator.py]: I know it's not your changes but can you resolve this Guard against a non-positive planning size, and reconcile the dxp_lx_frac_avail semantics with the docstring.

Two concerns:

- The new unconditional - lx_backend_spill_margin has no floor. For configs where (2<<20) * (1.0 - config.dxp_lx_frac_avail) is at or below 100 KB (i.e. dxp_lx_frac_avail ≳ 0.95), this returns 0 or a negative size, which is then handed to the layout solvers as the available LX budget. Clamp to a sane minimum.

- The docstring describes dxp_lx_frac_avail as "the frontend's nominal dxp_lx_frac_avail partition," but the sizing uses its complement (1.0 - config.dxp_lx_frac_avail). If dxp_lx_frac_avail is truly the frontend's share, the frontend budget should scale with config.dxp_lx_frac_avail, not 1.0 - config.dxp_lx_frac_avail. Please confirm which side of the partition this fraction denotes.

**Comment by shruthip-venkatesh** [comment]
Depends on https://github.com/torch-spyre/torch-spyre/pull/3289

File: torch_spyre/_inductor/decompositions.py (hunk 1 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/decompositions.py:1044]: Addressed in L1063

File: torch_spyre/_inductor/decompositions.py [hunk 2]
```diff
@@ -997,6 +997,89 @@ def spyre_prod_dim_int(
...
+    cols = self.shape[-1]
+    rows = self.numel() // cols
+    # Collapse the broadcast dim: one bool per row.
+    mask_row = mask[..., 0].reshape(rows)
+    source_2d = source.reshape(-1, cols)
+    # Empty source is legal when the mask is all-False (True-count == 0).
+    # shape[0] is static, so this is a compile-time branch: no data-dependent
+    # control flow. where(mask, _, self) would always return self anyway.
+    if source_2d.shape[0] == 0:
+        return self.clone()
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/decompositions.py]: Preserve the insufficient-source error contract.

When source_2d is empty and mask_row contains a selected row, Line 1064 silently returns self; native masked_scatter raises because the source is too short. More generally, selected row indices can exceed source_2d.shape[0] - 1. Add an equivalent runtime source-length check (or decline this lowering) before gathering, and cover empty/undersized sources with selected rows.

File: torch_spyre/_inductor/scratchpad/allocator.py [hunk 3]
```diff
@@ -734,6 +743,35 @@ def _op_short_name(op: Any) -> str:
...
+    ``uses`` is the sorted access list from ``calculate_liveness`` (writes and
+    reads), so the endpoints are the producer and last consumer; both are
+    already covered by the user check, hence the exclusive range.
+    """
+    if len(uses) <= 1:
+        return False
+    return any(
+        isinstance(graph.operations[i], ExternKernel)
+        for i in range(uses[0] + 1, uses[-1])
+    )
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/scratchpad/allocator.py]: Avoid rescanning entire live intervals per buffer.

This helper is called for every candidate buffer, so graphs with many long-lived buffers can incur O(buffers × operations) analysis time. Precompute sorted extern-kernel indices or a prefix count once in _residency_reasons, then answer each interval query in O(log E) or O(1).

File: torch_spyre/_inductor/scratchpad/allocator.py [hunk 4]
```diff
@@ -283,6 +283,15 @@ def _buffer_residency_reason(
...
             return "extern kernel user"
+        # An ExternKernel that *passes through* the buffer's live range
+        # (producer ... consumer) without accessing it still invalidates LX
+        # residency: extern ops launch as separate kernels (often CPU
+        # fallbacks) and LX is core-local scratch that does not survive the
+        # round-trip. The "extern kernel user" check above only catches an
+        # extern op that reads/writes the buffer; this catches one that
+        # merely executes between producer and consumer.
+        if _extern_kernel_in_live_range(graph, uses):
+            return "extern kernel in live range"
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/scratchpad/allocator.py]: _input_residency_reason can still return None for an input whose uses span an unrelated ExternKernel, allowing _build_bound_buffers to clone it into LX. That crosses the same kernel boundary this branch protects for computed buffers and can produce invalid results.

File: torch_spyre/_inductor/decompositions.py [hunk 5]
```diff
@@ -997,6 +997,89 @@ def spyre_prod_dim_int(
...
+    row_idx = (pos * mask_row.to(torch.float32)).to(torch.int64)
+    # The gather stays 2D (its args are indirect, so the pointwise dim_order
+    # projection skips them), but the `where` must stay at `self`'s rank:
+    # that projection computes `rank_diff = len(output) - len(arg)` and only
+    # handles inputs of *lower* rank. A lower-rank output against the full-rank
+    # ND mask gives rank_diff < 0, which shifts dims the wrong way and builds a
+    # layout whose dim_order rank no longer matches host_size ("Incompatible
+    # host_size and dim_order"). Reshaping back to ND keeps every `where`
+    # input rank-aligned with the output.
+    gathered = source_2d[row_idx].reshape(self.shape)
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/decompositions.py:1043]: https://github.com/torch-spyre/torch-spyre/pull/3308/changes#r3658457083

File: torch_spyre/_inductor/scratchpad/allocator.py
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/scratchpad/allocator.py]: Thanks for your review ! We will consider this as a separate task as it is not related to this PR fix.

File: torch_spyre/_inductor/decompositions.py (hunk 5 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/decompositions.py:1043]: https://github.com/torch-spyre/torch-spyre/pull/3308/changes#diff-11d87f25744b29772914fdaaaca93f22f94153b17f541201d422fe6e93ae7659R919

File: torch_spyre/_inductor/decompositions.py (hunk 2 above)
Comment (shruthip-venkatesh) [reply torch_spyre/_inductor/decompositions.py]: Addressed with the check `mask_row.sum() <= source_2d.shape[0],`

**Comment by shruthip-venkatesh** [comment]
@tardieu / @dgrove-oss / @avery-blanchard ,
Could you review this PR ?

**Comment by shruthip-venkatesh** [comment]
@tardieu  / @dgrove-oss  / @avery-blanchard  ,
Could you review this PR ?

File: docs/source/user_guide/examples/indirect_access/masked_scatter.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    expected = torch.masked_scatter(inp_rt, mask_nd_cpu, src_rt)
+
+    # Compile and run on device
+    got = torch.compile(_masked_scatter)(inp_dev, mask_nd_dev, src_dev).cpu()
+
+    # Check
+    n_diff = int((got != expected).sum())
+    total = got.numel()
+    if n_diff == 0:
+        print(f"PASSED  ({total} elements, {N_TRUE} selected rows)")
```
Comment (avery-blanchard) [inline docs/source/user_guide/examples/indirect_access/masked_scatter.py]: if this passes and runs with the changes in this PR, can we make it an actual test rather than an example in the docs?

Comment (shruthip-venkatesh) [reply docs/source/user_guide/examples/indirect_access/masked_scatter.py]: Yes, I have pushed the examples into unit test files along with few more scenarios covered.

**Review by avery-blanchard — APPROVED** [review]
looks good to me, thanks!
